docs/source/daml/intro/9_Functional101.rst

```diff
@@ -0,0 +1,498 @@
+  :language: daml
+  :start-after: -- INT_OR_TEXT_BEGIN
+  :end-before: -- INT_OR_TEXT_END
+
+Branching in Actions
+~~~~~~~~~~~~~~~~~~~~
+
+The most common case where this becomes important is inside ``do`` blocks. Say we want to create a contract of one type in one case, and of another type in another case. Let's say we have two template types and want to write a function that creates an ``S`` if a condition is met, and a ``T`` otherwise.
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- S_T_BEGIN
+  :end-before: -- S_T_END
+
+It would be tempting to write a simple ``if..else``, but it won't typecheck:
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- TYPE_ERROR2_BEGIN
+  :end-before: -- TYPE_ERROR2_END
+
+We have two options:
+
+1. Use the ``Either`` trick from above.
+2. Get rid of the return types.
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- S_OR_T_BEGIN
+  :end-before: -- S_OR_T_END
+
+The latter is so common that there is a utility function in ``DA.Action`` to get rid of the return type: ``void : Functor f => f a -> f ()``.
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- VOID_BEGIN
+  :end-before: -- VOID_END
+
+``void`` also helps express control flow of the type "Create a ``T`` only if a condition is met.
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- CUSTOM_WHEN_BEGIN
+  :end-before: -- CUSTOM_WHEN_END
+
+Note that we still need the ``else`` clause of the same type ``()``. This pattern is so common, it's encapsulated in the standard library function ``DA.Action.when : (Applicative f) => Bool -> f () -> f ()``. 
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- WHEN_BEGIN
+  :end-before: -- WHEN_END
+
+You've already seen this in use in Chapter 8:
+
+.. literalinclude:: daml/daml-intro-8-upgrade/daml/Test/Intro/Asset/Upgrade/V2.daml
+  :language: daml
+  :start-after: -- RUN_COMPLETE_UPGRADE_BEGIN
+  :end-before: -- RUN_COMPLETE_UPGRADE_END
+
+With ``case``, ``if..else``, ``void`` and ``when``, you can express all branching. However, one additional feature you may want to learn is guards. They are not covered here, but can help avoid deeply nested ``if..else`` blocks. Here's just one example. The Haskell sources at the beginning of the chapter cover this topic in more depth.
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- TELL_SIZE_BEGIN
+  :end-before: -- TELL_SIZE_END
+
+.. _looping:
+
+Looping
+.......
+
+Other than branching, the most common form of control flow is looping. Looping is usually used to iteratively modify some state. We'll use pseudocode in this section to illustrate the procedural way of doing things. 
+
+.. code-block:: JavaScript
+
+  function sum(intArr) {
+    int result = 0;
```

Comment:
Those examples seem like a somewhat weird mix of javascript and not quite javascript. Maybe just go for full javascript?

Reply:
But then I'd have to test that it's valid javascript 😢

Reply:
Ok, that was easier than I thought. It's now actual JavaScript